### PR TITLE
Move dxil.dll to DXC repo.

### LINKED ISSUE #6769: DXIL Validator Hash
This issue serves as a progress tracker for [DXIL Validator Hash](
https://github.com/microsoft/hlsl-specs/blob/main/proposals/infra/INF-0004-validator-hashing.md)

# Tasks

- [ ] Build the DXIL validator library both as a dynamic and static library to allow full flexibility.
- [ ] Open source the implementation of dxil.dll.
- [ ] Support bypassing Hash Verification.

### PR DESCRIPTION
**Pull request by python3kgae**
Build dxil.dll in DXC repo.
It will shares same validator code with dxcompiler.dll.

Fixes #6769

### CODE REVIEW

File: tools/clang/tools/dxcvalidator/CMakeLists.txt [hunk 1]
```diff
@@ -8,7 +8,9 @@ set( LLVM_LINK_COMPONENTS
   DXIL
   DxilContainer
   DxilHash
+  DxilRootSignature
   DxilValidation
+  HLSL       # for AreDxilResourcesDense and DxilSignatureAllocator which are used in DxilValidation
```
Comment (pow2clk) [inline tools/clang/tools/dxcvalidator/CMakeLists.txt:13]: Given that there are no code changes in this directory, I don't know why any changes would be needed here.

File: tools/clang/tools/dxildll/dxccontainerbuilder.cpp [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  // Call dxil.dll's containerbuilder
+  *V = nullptr;
+  CComPtr<DxcContainerBuilder> Result(
+      DxcContainerBuilder::Alloc(DxcGetThreadMallocNoRef()));
+  if (nullptr == Result.p)
+    return E_OUTOFMEMORY;
+
+  Result->Init();
+  return Result->QueryInterface(RRID, V);
+}
```
Comment (pow2clk) [inline tools/clang/tools/dxildll/dxccontainerbuilder.cpp:24]: I'll pass along the [feedback ](https://github.com/microsoft/DirectXShaderCompiler/pull/5952#discussion_r1382231012)I got from #5952 and suggest that this file be moved into dxildll.cpp

File: tools/clang/tools/dxildll/dxcvalidator.cpp [hunk 3]
```diff
@@ -0,0 +1,105 @@
+///////////////////////////////////////////////////////////////////////////////
+//                                                                           //
+// dxcvalidatorp.cpp                                                         //
```
Comment (pow2clk) [inline tools/clang/tools/dxildll/dxcvalidator.cpp:3]: nit: not what this file is called

File: tools/clang/tools/dxildll/dxcvalidator.cpp [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+#include "dxc/Support/WinIncludes.h"
+
+#include "dxc/Support/Global.h"
+#include "dxc/Support/microcom.h"
+#include "dxc/dxcapi.h"
+
+using namespace llvm;
+using namespace hlsl;
+
+class DxcValidator : public IDxcValidator2, public IDxcVersionInfo {
```
Comment (pow2clk) [inline tools/clang/tools/dxildll/dxcvalidator.cpp:23]: I'm trying to think of a reason why we might want to enable querying the git commit info from this and coming up with nothing, but if you can think of something, maybe we should inherit from `IDxcVersionInfo2`

File: tools/clang/tools/dxildll/dxildll.cpp [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+  HRESULT HR = (DxcInitThreadMalloc());
+  if (DXC_FAILED(HR))
+    goto Cleanup;
+  DxcSetThreadMallocToDefault();
+  MemSetup = true;
+  if (::llvm::sys::fs::SetupPerThreadFileSystem()) {
+    HR = E_FAIL;
+    goto Cleanup;
+  }
+Cleanup:
```
Comment (pow2clk) [inline tools/clang/tools/dxildll/dxildll.cpp:48]: nit: I know it's not your code originally, but this seems a flimsy excuse to use a goto as one of them just falls through anyway and the other could very easily be a conditional block
```suggestion
  if (!DXC_FAILED(HR)) {
    DxcSetThreadMallocToDefault();
    MemSetup = true;
    if (::llvm::sys::fs::SetupPerThreadFileSystem())
      HR = E_FAIL;
  }
```

**Review by pow2clk — COMMENTED** [review]
If not for the dxcvalidator/CMakeList.txt change I don't understand, I'd approve this. The other suggestions you can take or leave.

File: tools/clang/tools/dxildll/dxildll.cpp (hunk 5 above)
Comment (python3kgae) [reply tools/clang/tools/dxildll/dxildll.cpp]: Updated.

File: tools/clang/tools/dxildll/dxcvalidator.cpp (hunk 4 above)
Comment (python3kgae) [reply tools/clang/tools/dxildll/dxcvalidator.cpp:23]: Unable to find a reason as well.

File: tools/clang/tools/dxildll/dxcvalidator.cpp (hunk 3 above)
Comment (python3kgae) [reply tools/clang/tools/dxildll/dxcvalidator.cpp]: Updated.

File: tools/clang/tools/dxildll/dxccontainerbuilder.cpp (hunk 2 above)
Comment (python3kgae) [reply tools/clang/tools/dxildll/dxccontainerbuilder.cpp]: Updated.

File: tools/clang/tools/dxcvalidator/CMakeLists.txt (hunk 1 above)
Comment (python3kgae) [reply tools/clang/tools/dxcvalidator/CMakeLists.txt]: Moved to CMakeLists.txt for dxildll.

Comment (python3kgae) [reply tools/clang/tools/dxcvalidator/CMakeLists.txt]: But gcc complains after remove change here.

Comment (pow2clk) [reply tools/clang/tools/dxcvalidator/CMakeLists.txt]: Okay. I suspect I know what's going on here. I think that these dependencies were always there, but they were satisfied indirectly by other libraries by chance. The linux linker is really particular that way.

**Comment by pow2clk** [comment]
GCC is still failing. Is the move of the container builder into dxildll.cpp the cause? I remember something like that with my similar change.

File: tools/clang/test/CMakeLists.txt [hunk 6]
```diff
@@ -77,7 +77,11 @@ endif()
 set(CLANG_TEST_DEPS dxc dxa dxopt dxl dxv dxr dxcompiler clang-tblgen llvm-config opt FileCheck count not ClangUnitTests)
 if (WIN32)
 list(APPEND CLANG_TEST_DEPS
-     dxc_batch ExecHLSLTests
+     dxc_batch ExecHLSLTests dxildll
+    )
+elseif(NOT ${CMAKE_SYSTEM_NAME} MATCHES "Darwin")
```
Comment (llvm-beanz) [inline tools/clang/test/CMakeLists.txt:82]: I'm confused by this. Why is this needed?

Comment (python3kgae) [reply tools/clang/test/CMakeLists.txt:82]: Shall we enable libdxil to mac?
We only release dxildll for linux and windows now.

Comment (llvm-beanz) [reply tools/clang/test/CMakeLists.txt:82]: I think one of the biggest benefits of us open sourcing this is to allow people to build it on macOS where we don't currently have supported binary releases.

File: tools/clang/tools/dxcvalidator/CMakeLists.txt [hunk 7]
```diff
@@ -9,6 +9,8 @@ set( LLVM_LINK_COMPONENTS
   DxilContainer
   DxilHash
   DxilValidation
+  DxilRootSignature
+  HLSL       # for gcc link error on AreDxilResourcesDense and DxilSignatureAllocator which are used in DxilValidation
```
Comment (tex3d) [inline tools/clang/tools/dxcvalidator/CMakeLists.txt:13]: If these are the only dependencies on `HLSL.lib`, I think we could avoid this dependency with two things:
- `AreDxilResourcesDense`: this is only used to verify that resource IDs in metadata lists are sequential, starting at 0.  This could be locally checked in a simpler way for validation purposes.
- `DxilSignatureAllocator`: this is already implemented in a `.inl` file to make it easy to separate this dependency.

File: tools/clang/tools/dxildll/dxcvalidator.cpp (hunk 4 above)
Comment (tex3d) [reply tools/clang/tools/dxildll/dxcvalidator.cpp:23]: The DXIL major/minor version info is not sufficient to identify different validator versions within the same release cycle.  Since this change introduces a build of `dxil.dll` into every build of the public DXC project, it seems more important than before to retain the additional version information going forward.

What do you think?

Comment (damyanp) [reply tools/clang/tools/dxildll/dxcvalidator.cpp:23]: > it seems more important than before to retain the additional version information going forward.

Totally agree with this.  I hope the answer isn't to use a commit hash though.  Ideally we'd get some kind minor-minor version number we can increment as appropriate? Maybe this is something we'll need to consider independently of this though

File: tools/clang/tools/dxcvalidator/CMakeLists.txt (hunk 7 above)
Comment (python3kgae) [reply tools/clang/tools/dxcvalidator/CMakeLists.txt:13]: AreDxilResourceDense should be easy to move to validation.
Where should we put DxilSignatureAllocator?
It is used in Validation and DxilPackSignatureElement.

Comment (tex3d) [reply tools/clang/tools/dxcvalidator/CMakeLists.txt:13]: One option for `DxilSignatureAllocator.inl` would be to include it in a `.cpp` file under `dxildll` to provide an implementation when not linking in HLSL.lib.  Another solution might be to move the DxilPackSignatureElement stuff into the LLVMDXIL lib, or if we don't want to add it there, we could put it into a new lib instead.

Comment (python3kgae) [reply tools/clang/tools/dxcvalidator/CMakeLists.txt:13]: Created issue https://github.com/microsoft/DirectXShaderCompiler/issues/6898 to track this.

**Review by pow2clk — APPROVED** [review]
Thanks!

File: tools/clang/tools/dxildll/dxcvalidator.cpp (hunk 4 above)
Comment (python3kgae) [reply tools/clang/tools/dxildll/dxcvalidator.cpp]: We already have version number in dxc --version like
`dxcompiler.dll: 1.8 - 1.8.0.4710 (test_gcc_bld, 913ae2ce8); dxil.dll: 1.8(1.8.0.4710)`.
